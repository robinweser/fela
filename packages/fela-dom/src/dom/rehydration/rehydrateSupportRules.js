/* @flow */
import extractSupportQuery from './extractSupportQuery'
import rehydrateRules from './rehydrateRules'

const SUPPORT_REGEX = /@supports[^{]+\{([\s\S]+?})\s*}/gi

export default function rehydrateSupportRules(
  css: string,
  media?: string = '',
  cache?: Object = {},
  specificityPrefix?: string = ''
): Object {
  let decl

  // eslint-disable-next-line no-cond-assign
  while ((decl = SUPPORT_REGEX.exec(css))) {
    const [ruleSet, cssRules] = decl

    const supportQuery = extractSupportQuery(ruleSet)
    rehydrateRules(cssRules, media, supportQuery, cache, specificityPrefix)
  }

  return cache
}
