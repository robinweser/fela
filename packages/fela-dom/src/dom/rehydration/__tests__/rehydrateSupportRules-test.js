import rehydrateSupportRules from '../rehydrateSupportRules'

describe('Rehydrating @supports rules', () => {
  it('should rehydrate the renderer cache', () => {
    expect(
      rehydrateSupportRules(
        '@supports(display:grid){.a{color:red}.b{color:blue}}@supports(display:grid) and (display:flex){.c{color:green}}'
      )
    ).toMatchSnapshot()
  })
  it('should rehydrate the renderer cache with specifityPrefix', () => {
    expect(
      rehydrateSupportRules(
        '@supports(display:grid){.parentClass .a{color:red}.parentClass .b{color:blue}}@supports(display:grid) and (display:flex){.parentClass .c{color:green}}',
        '',
        {},
        '.parentClass '
      )
    ).toMatchSnapshot()
  })
})
