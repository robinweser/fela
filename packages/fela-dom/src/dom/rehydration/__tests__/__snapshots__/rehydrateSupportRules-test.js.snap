// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rehydrating @supports rules should rehydrate the renderer cache 1`] = `
Object {
  "(display:grid) and (display:flex)colorgreen": Object {
    "className": "c",
    "declaration": "color:green",
    "media": "",
    "pseudo": "",
    "selector": ".c",
    "support": "(display:grid) and (display:flex)",
    "type": "RULE",
  },
  "(display:grid)colorblue": Object {
    "className": "b",
    "declaration": "color:blue",
    "media": "",
    "pseudo": "",
    "selector": ".b",
    "support": "(display:grid)",
    "type": "RULE",
  },
  "(display:grid)colorred": Object {
    "className": "a",
    "declaration": "color:red",
    "media": "",
    "pseudo": "",
    "selector": ".a",
    "support": "(display:grid)",
    "type": "RULE",
  },
}
`;

exports[`Rehydrating @supports rules should rehydrate the renderer cache with specifityPrefix 1`] = `
Object {
  "(display:grid) and (display:flex)colorgreen": Object {
    "className": "c",
    "declaration": "color:green",
    "media": "",
    "pseudo": "",
    "selector": ".parentClass .c",
    "support": "(display:grid) and (display:flex)",
    "type": "RULE",
  },
  "(display:grid)colorblue": Object {
    "className": "b",
    "declaration": "color:blue",
    "media": "",
    "pseudo": "",
    "selector": ".parentClass .b",
    "support": "(display:grid)",
    "type": "RULE",
  },
  "(display:grid)colorred": Object {
    "className": "a",
    "declaration": "color:red",
    "media": "",
    "pseudo": "",
    "selector": ".parentClass .a",
    "support": "(display:grid)",
    "type": "RULE",
  },
}
`;
